BiometricLoginKotlin/app/src/main/java/com/example/biometricloginsample/CryptographyManager.kt

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (C) 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+
+package com.example.biometricloginsample
+
+import android.content.Context
+import android.security.keystore.KeyGenParameterSpec
+import android.security.keystore.KeyProperties
+import com.google.gson.Gson
+import java.nio.charset.Charset
+import java.security.KeyStore
+import javax.crypto.Cipher
+import javax.crypto.KeyGenerator
+import javax.crypto.SecretKey
+import javax.crypto.spec.GCMParameterSpec
+
+interface CryptographyManager {
+
+    /**
+     * This method first gets or generates an instance of SecretKey and then initializes the Cipher
+     * with the key. The secret key uses [ENCRYPT_MODE][Cipher.ENCRYPT_MODE] is used.
+     */
+    fun getInitializedCipherForEncryption(keyName: String): Cipher
+
+    /**
+     * This method first gets or generates an instance of SecretKey and then initializes the Cipher
+     * with the key. The secret key uses [DECRYPT_MODE][Cipher.DECRYPT_MODE] is used.
+     */
+    fun getInitializedCipherForDecryption(keyName: String, initializationVector: ByteArray): Cipher
+
+    /**
+     * The Cipher created with [getInitializedCipherForEncryption] is used here
+     */
+    fun encryptData(plaintext: String, cipher: Cipher): CiphertextWrapper
+
+    /**
+     * The Cipher created with [getInitializedCipherForDecryption] is used here
+     */
+    fun decryptData(ciphertext: ByteArray, cipher: Cipher): String
+
+    fun persistCiphertextWrapperToSharedPrefs(
+        ciphertextWrapper: CiphertextWrapper,
+        context: Context,
+        filename: String,
+        mode: Int,
+        prefKey: String
+    )
+
+    fun getCiphertextWrapperFromSharedPrefs(
+        context: Context,
+        filename: String,
+        mode: Int,
+        prefKey: String
+    ): CiphertextWrapper?
+
+}
+
+fun CryptographyManager(): CryptographyManager = CryptographyManagerImpl()
```

Comment:
You're missing a verb here.

Reply:
Please elaborate

Reply:
That's intentional. It's a free function that acts like a public constructor. Hence the create- prefix is superfluous in this case.

Reply:
I have not seen this before. Please use `newInstance` instead.

Reply:
"I have not seen this before. " That doesn't make it bad :) . Do you find it confusing?

Reply:
It's not common to do this. Use newInstance instead.

Reply:
could you show the `newInstance` method?

Reply:
```suggestion
fun newInstance(): CryptographyManager = CryptographyManagerImpl()
```

Reply:
Thanks for the discussion; I very much appreciate it. But grammatically speaking, changing the name to `newInstance` would result in a dangling function. Also, as it turns out, I am in good company as `Job` in coroutine uses the same pattern: `public fun Job(parent: Job? = null): Job = JobImpl(parent)`.

Reply:
I agree on not having a dangling function here. Usually a newInstance function is a static function of the class it instantiates. So a `CryptographyManager.newInstance` would be the appropriate function.

Reply:
Or even instantiating `CryptographyManagerImpl` using a regular constructor.

Reply:
Given that the current usage is actually used by the Kotlin language architects themselves, and the solution you suggested was dangling and so wouldn't scale, would you be ok resolving this thread?